README
========================
|                      |
| GPUdb JavaScript API |
|                      |
========================


This repository contains the 4.2.0 version of the client-side API for GPUdb for
JavaScript.  Using this API requires rpm packages 'nodejs' and 'npm', the nodejs
package manager.

The repository also contains an example project.  To see it in action, load
example.html in any browswer and look at the console output in the developer
tools.



The documentation can be found at http://www.gpudb.com/docs/index.html.  The
JavaScript specific documentation can be found at:

* http://www.gpudb.com/docs/user/js_guide.html
* http://www.gpudb.com/docs/api/javascript/GPUdb.html


For the Node.js module, the following rpm packages are required:

* nodejs
* npm (the nodejs package manager)


CHANGELOG
========================
|                      |
| GPUdb JavaScript API |
|                      |
========================

Version 5.0.0:
----------------

* Updated endpoints


Version 4.2.0:
----------------

* Added Node.js module
* Updated endpoints


example/nodejs-example.js
var GPUdb = require("../nodejs/GPUdb.js");

console.log("Establishing a connection with GPUdb...");
var gpudb = new GPUdb("http://localhost:9191");

var operation_number = 0;

/*
 * A mechanism for making sure that the previous operation
 * performed has completed before starting the current operation.
 */
var next_operation = function() {
    if (operation_number < operations.length) {
        operations[operation_number++]();
    }
}


var build_callback = function(success, error) {
    return function(err, response) {
        if (err === null) {
            if (success !== undefined) {
                success(response);
            }

            next_operation();
        } else {
            if (error !== undefined) {
                error(err);
                next_operation();
            } else {
                console.log(err);
            }
        }
    };
}

// Need global scoping for the type ID and table name
var type_id;
var table_name = "my_table";

var operations = [
    // Clear all tables from the database
    function() {
        gpudb.clear_table("", null, {}, build_callback());
    },

    function() {
        gpudb.show_table("", {}, build_callback(function(response) {
            console.log(response);
        }));
    },

    // Register the data type for the table with GPUdb and get the type's ID
    function() {
        var my_type = new GPUdb.Type("my_type",
                new GPUdb.Type.Column("col1", "double"),
                new GPUdb.Type.Column("col2", "string"),
                new GPUdb.Type.Column("group_id", "string"));

        my_type.create(gpudb, build_callback( function( response ) {
            type_id = response;
        } ));
    },

    // Create the table
    function() {
        gpudb.create_table( table_name, type_id, {}, build_callback() );
    },

    // Generate the records to be inserted and insert them
    function() {
        var records = [];

        for (var i = 0; i < 10; i++) {
            records.push({
                col1: i + 0.1,
                col2: "string " + i,
                group_id: "Group 1"
            });
        }

	var insert_options = { "return_record_ids" : "true" }
        gpudb.insert_records( table_name, records, insert_options, build_callback(function(response) {
            console.log("Record IDs for newly inserted records: " + response.record_ids);
        }));
    },

    // Fetch the records from the table
    function() {
        gpudb.get_records( table_name, 0, -9999, {}, build_callback(function(response) {
            console.log("Retrieved records: ");
            console.log(response.data);
        }));
    },

    // Perform a filter operation on the table
    function() {
	var view_name = "view_1";
	var expression = ("col1 = 1.1");
        gpudb.filter( table_name, view_name, expression, {}, build_callback(function(response) {
            console.log("Number of filtered records: " + response.count);
        }));
    },

    // Fetch the records from the view (like reading from a regular table)
    function() {
        gpudb.get_records("view_1", 0, -9999, {}, build_callback(function(response) {
            console.log("Filtered records: ");
            console.log(response.data);
        }));
    },

    // Drop the view
    function() {
        gpudb.clear_table("view_1", null, {}, build_callback());
    },

    // Perform a filter operation on the table on two column_names
    function() {
        gpudb.filter( table_name, "view_1", "col1 <= 9 and group_id = 'Group 1'", {}, build_callback(function(response) {
            console.log("Number of records filtered by the second expression: " + response.count);
        }));
    },

    // Fetch the records from the view
    function() {
        gpudb.get_records("view_1", 0, -9999, {}, build_callback(function(response) {
            console.log("Second set of filtered records: ");
            console.log(response.data);
        }));
    },

    // Perform a filter by list operation
    function() {
	var view_name = "view_2";
	var column_values_map = {
	    col1 : [ "1.1", "2.1", "5.1" ]
	};
        gpudb.filter_by_list( table_name, view_name, column_values_map, {}, build_callback(function(response) {
            console.log("Number of records filtered by list: " + response.count);
        }));
    },

    // Fetch the records from the second view
    function() {
        gpudb.get_records("view_2", 0, -9999, {}, build_callback(function(response) {
            console.log("Records filtered by a list: ");
            console.log(response.data);
        }));
    },

    // Perform a filter by range operation
    function() {
        gpudb.filter_by_range( table_name, "view_3", "col1", 1, 5, {}, build_callback(function(response) {
            console.log("Number of records filtered by range: " + response.count);
        }));
    },

    // Fetch the records from the third view
    function() {
        gpudb.get_records("view_3", 0, -9999, {}, build_callback(function(response) {
            console.log("Records filtered by range: ");
            console.log(response.data);
        }));
    },

    // Perform an aggregate operation (statistics: sum, mean, count)
    function() {
        gpudb.aggregate_statistics( table_name, "col1", "sum,mean,count", {}, build_callback(function(response) {
            console.log("Statistics of values in 'col1': " + JSON.stringify(response.stats));
        }));
    },

    // Insert some more records
    function() {
        console.log("Inserting more records into the table...");
        var records = [];

        for (var i = 1; i < 8; i++) {
            records.push({
                col1: i + 10.1,
                col2: "string " + i,
                group_id: "Group 2"
            });
        }

        gpudb.insert_records( table_name, records, {}, build_callback());
    },

    // Find all unique values of a given column
    function() {
        gpudb.aggregate_unique( table_name, "group_id", 0, -9999, {}, build_callback(function(response) {
            console.log("Unique values in 'group_id': ");
            console.log(response.data);
        }));
    },

    // Aggregate values of a given column by grouping by its values
    function() {
        var column_names = [ "col2" ];
        gpudb.aggregate_group_by( table_name, column_names, 0, -9999, {}, build_callback(function(response) {
            console.log("Group by results: ");
            console.log(response.data);
        }));
    },

    // Second group by
    function() {
        var column_names = [ "group_id", "count(*)", "sum(col1)", "avg(col1)" ];
        gpudb.aggregate_group_by( table_name, column_names, 0, -9999, {}, build_callback(function(response) {
            console.log("Second group by results: ");
            console.log(response.data);
        }));
    },

    // Third group by
    function() {
        gpudb.aggregate_group_by( table_name, [ "group_id", "sum(col1*10)" ], 0, -9999, {}, build_callback(function(response) {
            console.log("Third group by results: ");
            console.log(response.data);
        }));
    },

    // Insert some more records
    function() {
        console.log("Inserting more records into the table...");
        var records = [];

        for (var i = 4; i < 10; i++) {
            records.push({
                col1: i + 0.6,
                col2: "string 2" + i,
                group_id: "Group 1"
            });
        }

        gpudb.insert_records( table_name, records, {}, build_callback());
    },

    // Perform a histogram calculation
    function() {
	var start = 1.1;
	var end = 2;
	var interval = 1;
        gpudb.aggregate_histogram( table_name, "col1", start, end, interval, {}, build_callback(function(response) {
            console.log("Histogram results: ");
            console.log(response);
        }));
    },

    // Drop the original table (will automatically drop all views of it)
    function() {
        gpudb.clear_table( table_name, null, {}, build_callback());
    },

    // Check that no view of that table is available anymore.
    function() {
        gpudb.show_table("view_3", {}, build_callback(function(response) {
            console.log("Should not get here!");
        }, function(error) {
            console.log("View 'view_3' not available as expected.");
        }));
    }
];

next_operation();




/*
Expected Output:
================

Establishing a connection with GPUdb...
{ table_name: '',
  table_names: [],
  is_collection: [],
  is_view: [],
  type_ids: [],
  type_schemas: [],
  type_labels: [],
  properties: [],
  ttls: [],
  sizes: [],
  full_sizes: [],
  total_size: -1,
  total_full_size: -1 }
Record IDs for newly inserted records: 0010300000000000_0000000000000000,0010300000000000_0000000000000001,0010300000000000_0000000000000002,0010300000000000_0000000000000003,0010300000000000_0000000000000004,0010300000000000_0000000000000005,0010300000000000_0000000000000006,0010300000000000_0000000000000007,0010300000000000_0000000000000008,0010300000000000_0000000000000009
Retrieved records: 
[ { col1: 0.1, col2: 'string 0', group_id: 'Group 1' },
  { col1: 1.1, col2: 'string 1', group_id: 'Group 1' },
  { col1: 2.1, col2: 'string 2', group_id: 'Group 1' },
  { col1: 3.1, col2: 'string 3', group_id: 'Group 1' },
  { col1: 4.1, col2: 'string 4', group_id: 'Group 1' },
  { col1: 5.1, col2: 'string 5', group_id: 'Group 1' },
  { col1: 6.1, col2: 'string 6', group_id: 'Group 1' },
  { col1: 7.1, col2: 'string 7', group_id: 'Group 1' },
  { col1: 8.1, col2: 'string 8', group_id: 'Group 1' },
  { col1: 9.1, col2: 'string 9', group_id: 'Group 1' } ]
Number of filtered records: 1
Filtered records: 
[ { col1: 1.1, col2: 'string 1', group_id: 'Group 1' } ]
Number of records filtered by the second expression: 9
Second set of filtered records: 
[ { col1: 0.1, col2: 'string 0', group_id: 'Group 1' },
  { col1: 1.1, col2: 'string 1', group_id: 'Group 1' },
  { col1: 2.1, col2: 'string 2', group_id: 'Group 1' },
  { col1: 3.1, col2: 'string 3', group_id: 'Group 1' },
  { col1: 4.1, col2: 'string 4', group_id: 'Group 1' },
  { col1: 5.1, col2: 'string 5', group_id: 'Group 1' },
  { col1: 6.1, col2: 'string 6', group_id: 'Group 1' },
  { col1: 7.1, col2: 'string 7', group_id: 'Group 1' },
  { col1: 8.1, col2: 'string 8', group_id: 'Group 1' } ]
Number of records filtered by list: 3
Records filtered by a list: 
[ { col1: 1.1, col2: 'string 1', group_id: 'Group 1' },
  { col1: 2.1, col2: 'string 2', group_id: 'Group 1' },
  { col1: 5.1, col2: 'string 5', group_id: 'Group 1' } ]
Number of records filtered by range: 4
Records filtered by range: 
[ { col1: 1.1, col2: 'string 1', group_id: 'Group 1' },
  { col1: 2.1, col2: 'string 2', group_id: 'Group 1' },
  { col1: 3.1, col2: 'string 3', group_id: 'Group 1' },
  { col1: 4.1, col2: 'string 4', group_id: 'Group 1' } ]
Statistics of values in 'col1': {"count":10,"mean":4.6,"sum":46}
Inserting more records into the table...
Unique values in 'group_id': 
{ column_1: [ 'Group 1', 'Group 2' ],
  column_headers: [ 'group_id' ] }
Group by results: 
{ column_1: 
   [ 'string 0',
     'string 1',
     'string 2',
     'string 3',
     'string 4',
     'string 5',
     'string 6',
     'string 7',
     'string 8',
     'string 9' ],
  column_2: [ 1, 2, 2, 2, 2, 2, 2, 2, 1, 1 ],
  column_headers: [ 'col2', 'count(*)' ] }
Second group by results: 
{ column_1: [ 'Group 1', 'Group 2' ],
  column_2: [ 10, 7 ],
  column_3: [ 46, 98.69999999999999 ],
  column_4: [ 4.6, 14.1 ],
  column_headers: [ 'group_id', 'count(*)', 'sum(col1)', 'avg(col1)' ] }
Third group by results: 
{ column_1: [ 'Group 1', 'Group 2' ],
  column_2: [ 460, 987 ],
  column_headers: [ 'group_id', 'sum(col1*10)' ] }
Inserting more records into the table...
Histogram results: 
{ counts: [ 1 ], start: 1.1, end: 2 }
View 'view_3' not available as expected.

*/



example/example.js
'use strict';


/*
The script is called from an HTML file (opened in a browser) with the following content:

<!DOCTYPE html>
<html>
<head>
</head>

<body>
<script language="javascript" src="../GPUdb.js"> </script>
<script language="javascript" src="example.js"> </script>
</body>

</html>

*/


// Execute the 'main' function
main();


function main()
{
    console.log( "Establishing a connection with GPUdb..." );
    var gpudb = new GPUdb( "http://localhost:9191" );

    // Clear all tables from the database
    gpudb.clear_table( "" );

    var show_table_rsp = gpudb.show_table( "", {} );
    console.log( JSON.stringify( show_table_rsp ) );

    // Declare the data type for the table
    var my_type = {
        "type": "record",
	"name": "my_type_1",
	"fields": [
	    {"name":"col1","type":"double"},
	    {"name":"col2","type":"string"},
	    {"name":"group_id","type":"string"}
	]
    };
    // Convenient variables for referring to the columns
    var col1 = "col1";
    var col2 = "col2";
    var group_id = "group_id";

    // Register the data type with GPUdb and get the type's ID
    var create_type_rsp = gpudb.create_type( JSON.stringify( my_type ), "my_type" );
    var type_id = create_type_rsp.type_id;

    // Create a table
    var table_name = "my_table";
    var create_table_rsp = gpudb.create_table( table_name, type_id );

    // Generate records to be inserted 
    var records = [];
    for (var i = 0; i < 10; i++)
    {
	var record = {
	    col1 : (i + 0.1),
	    col2 : ("string " + i),
	    group_id : "Group 1"
	};
	records.push( record );
    }

    // This option will return IDs per record with which we can refer
    // to particular records later as needed
    var insert_options = { "return_record_ids" : "true" }
    var insert_records_rsp = gpudb.insert_records( table_name, records,
						   insert_options );
    console.log( "Record IDs for newly inserted records: "
    		 + insert_records_rsp["record_ids"] );

    // Fetch the records from the table
    var get_records_rsp = gpudb.get_records( table_name );
    console.log( "Retrieved records: " );
    console.log( JSON.stringify( get_records_rsp["data"] ) );

    // Perform a filter operation on the table
    var view_name_1 = "view_1";
    var expression = (col1 + " = 1.1");
    var filter_rsp = gpudb.filter( table_name, view_name_1, expression );
    console.log( "Number of filtered records: " + filter_rsp["count"] );

    // Fetch the records from the view (like reading from a regular table)
    var get_records_rsp = gpudb.get_records( view_name_1 );
    console.log( "Filtered records: " );
    console.log( JSON.stringify( get_records_rsp["data"] ) );

    // Drop the view
    gpudb.clear_table( view_name_1 );

    // Perform a filter operation on the table on two columns
    expression = (col1 + " <= 9 and " + group_id + '= "Group 1"');
    filter_rsp = gpudb.filter( table_name, view_name_1, expression );
    console.log( "Number of records filtered by the second expression: "
		 + filter_rsp["count"] );

    // Fetch the records from the view
    get_records_rsp = gpudb.get_records( view_name_1 );
    console.log( "Second set of filtered records: " );
    console.log( JSON.stringify( get_records_rsp["data"] ) );

    // Perform a filter by list operation
    var view_name_2 = "view_2";
    var column_values_map = {
	col1 : [ "1.1", "2.1", "5.1" ]
    };
    var filter_by_list_rsp = gpudb.filter_by_list( table_name, view_name_2,
						   column_values_map );
    console.log( "Number of records filtered by list: "
		 + filter_by_list_rsp["count"] );

    // Fetch the records from the second view
    get_records_rsp = gpudb.get_records( view_name_2 );
    console.log( "Records filtered by a list: " );
    console.log( JSON.stringify( get_records_rsp["data"] ) );

    // Perform a filter by range operation
    var view_name_3 = "view_3";
    var lower_bound = 1;
    var upper_bound = 5;
    var filter_by_range_rsp = gpudb.filter_by_range( table_name,
						     view_name_3,
						     col1,
						     lower_bound,
						     upper_bound );
    console.log( "Number of records filtered by range: "
		 + filter_by_range_rsp["count"] );

    // Fetch the records from the second view
    get_records_rsp = gpudb.get_records( view_name_3 );
    console.log( "Records filtered by a range: " );
    console.log( JSON.stringify( get_records_rsp["data"] ) );


    // Perform an aggregate operation (statistics: sum, mean, count)
    var statistics = "sum,mean,count"
    var stats_rsp = gpudb.aggregate_statistics( table_name, col1, statistics )
    console.log( "Statistics of values in 'col1': " + stats_rsp['stats'] );



    // Insert some more records
    console.log( "Inserting more records into the table..." );
    var records = [];
    for (var i = 1; i < 8; i++)
    {
	var record = {
	    col1 : (i + 10.1),
	    col2 : ("string " + i),
	    group_id : "Group 2" // unique from the first group of records
	};
	records.push( record );
    }
    gpudb.insert_records( table_name, records );


    // Find all unique values of a given column
    var unique_rsp = gpudb.aggregate_unique( table_name, group_id, 0 )
    console.log( "Unique of values in 'group_id': " );
    console.log( JSON.stringify( unique_rsp['data'] ) );

    // Aggregate values of a given column by grouping by its values
    var column_names = [ col2 ];
    var group_by_rsp = gpudb.aggregate_group_by( table_name, column_names, 0 )
    console.log( "Group by results: " );
    console.log( JSON.stringify( group_by_rsp['data'] ) );

    // Second group by
    var column_names = [ group_id, "count(*)", "sum(col1)", "avg(col1)" ];
    group_by_rsp = gpudb.aggregate_group_by( table_name, column_names, 0 )
    console.log( "Second group by results: " );
    console.log( JSON.stringify( group_by_rsp['data'] ) );

    // Third group by
    var column_names = [ group_id, "sum(col1*10)" ];
    group_by_rsp = gpudb.aggregate_group_by( table_name, column_names, 0 )
    console.log( "Third group by results: " );
    console.log( JSON.stringify( group_by_rsp['data'] ) );


    // Insert some more records
    console.log( "Inserting more records into the table..." );
    var records = [];
    for (var i = 4; i < 10; i++)
    {
	var record = {
	    col1 : (i + 0.6),
	    col2 : ("string 2" + i),
	    group_id : "Group 1"
	};
	records.push( record );
    }
    gpudb.insert_records( table_name, records );


    // Perform a histogram calculation
    var start = 1.1;
    var end = 2;
    var interval = 1;
    var histogram_rsp = gpudb.aggregate_histogram( table_name, col1,
						   start, end,
						   interval )
    console.log( "Histogram results: " );
    console.log( JSON.stringify( histogram_rsp ) );


    // Drop the original table (will automatically drop all views of it)
    gpudb.clear_table( table_name );

    // Check that no view of that table is available anymore.
    // Using a callback function to check the error status of the query.
    gpudb.show_table( view_name_3, "",
		      function( err, data ) // callback function
		      {
			  if (err !== null)
			  {
			      console.log( "View '" + view_name_3 + "' not available as expected." );
			  }
		      } );

};





/*
Expected Output:
================

Establishing a connection with GPUdb...

example.js:17 {"table_name":"","table_names":[],"is_collection":[],"is_view":[],"type_ids":[],"type_schemas":[],"type_labels":[],"properties":[],"ttls":[],"sizes":[],"full_sizes":[],"total_size":-1,"total_full_size":-1}

example.js:59 Record IDs for newly inserted records: 0020000002700000_0000000000000000,0020000002700000_0000000000000001,0020000002700000_0000000000000002,0020000002700000_0000000000000003,0020000002700000_0000000000000004,0020000002700000_0000000000000005,0020000002700000_0000000000000006,0020000002700000_0000000000000007,0020000002700000_0000000000000008,0020000002700000_0000000000000009

example.js:64 Retrieved records: 

example.js:65 [{"col1":0.1,"col2":"string 0","group_id":"Group 1"},{"col1":1.1,"col2":"string 1","group_id":"Group 1"},{"col1":2.1,"col2":"string 2","group_id":"Group 1"},{"col1":3.1,"col2":"string 3","group_id":"Group 1"},{"col1":4.1,"col2":"string 4","group_id":"Group 1"},{"col1":5.1,"col2":"string 5","group_id":"Group 1"},{"col1":6.1,"col2":"string 6","group_id":"Group 1"},{"col1":7.1,"col2":"string 7","group_id":"Group 1"},{"col1":8.1,"col2":"string 8","group_id":"Group 1"},{"col1":9.1,"col2":"string 9","group_id":"Group 1"}]

example.js:71 Number of filtered records: 1

example.js:75 Filtered records: 

example.js:76 [{"col1":1.1,"col2":"string 1","group_id":"Group 1"}]

example.js:84 Number of records filtered by the second expression: 9

example.js:89 Second set of filtered records: 

example.js:90 [{"col1":0.1,"col2":"string 0","group_id":"Group 1"},{"col1":1.1,"col2":"string 1","group_id":"Group 1"},{"col1":2.1,"col2":"string 2","group_id":"Group 1"},{"col1":3.1,"col2":"string 3","group_id":"Group 1"},{"col1":4.1,"col2":"string 4","group_id":"Group 1"},{"col1":5.1,"col2":"string 5","group_id":"Group 1"},{"col1":6.1,"col2":"string 6","group_id":"Group 1"},{"col1":7.1,"col2":"string 7","group_id":"Group 1"},{"col1":8.1,"col2":"string 8","group_id":"Group 1"}]

example.js:99 Number of records filtered by list: 3

example.js:104 Records filtered by a list: 

example.js:105 [{"col1":1.1,"col2":"string 1","group_id":"Group 1"},{"col1":2.1,"col2":"string 2","group_id":"Group 1"},{"col1":5.1,"col2":"string 5","group_id":"Group 1"}]

example.js:116 Number of records filtered by range: 4

example.js:121 Records filtered by a range: 

example.js:122 [{"col1":1.1,"col2":"string 1","group_id":"Group 1"},{"col1":2.1,"col2":"string 2","group_id":"Group 1"},{"col1":3.1,"col2":"string 3","group_id":"Group 1"},{"col1":4.1,"col2":"string 4","group_id":"Group 1"}]

example.js:128 Statistics of values in 'col1': [object Object]

example.js:133 Inserting more records into the table...

example.js:149 Unique of values in 'group_id': 

example.js:150 {"column_1":["Group 1","Group 2"],"column_headers":["group_id"]}

example.js:155 Group by results: 

example.js:156 {"column_1":["string 0","string 1","string 2","string 3","string 4","string 5","string 6","string 7","string 8","string 9"],"column_2":[1,2,2,2,2,2,2,2,1,1],"column_headers":["col2","count(*)"]}

example.js:161 Second group by results: 

example.js:162 {"column_1":["Group 1","Group 2"],"column_2":[10,7],"column_3":[46,98.69999999999999],"column_4":[4.6,14.1],"column_headers":["group_id","count(*)","sum(col1)","avg(col1)"]}

example.js:167 Third group by results: 

example.js:168 {"column_1":["Group 1","Group 2"],"column_2":[460,987],"column_headers":["group_id","sum(col1*10)"]}

example.js:172 Inserting more records into the table...

example.js:193 Histogram results: 

example.js:194 {"counts":[1],"start":1.1,"end":2}

example.js:206 View 'view_3' not available as expected.

*/
